enhancements/20240523-support-rwx-volume-to-downstream-cluster.md

```diff
@@ -0,0 +1,110 @@
+
+## Summary
+
+One of the harvester's classical user scenarios is the [downstream cluster](https://docs.harvesterhci.io/v1.3/rancher/rancher-integration). We should support the RWX volume for various applications running on it.
+
+### Related Issues
+
+https://github.com/harvester/harvester/issues/1992
+
+## Motivation
+
+Currently, the downstream cluster only supports the RWO volume. However, several applications need the RWX volume to work. Because Longhorn supports the RWX volume, we could merge the Longhorn RWX volume, and then users can create an RWX volume with their application.
+
+### Goals
+
+Make sure the downstream cluster can create workloads with the RWX volume.
+
+### Non-goals [optional]
+
+Create an endpoint for Harvester VM to use to have the shared filesystem.
+
+## Proposal
+
+The basic idea is to use the Longhorn RWX volume. It provides the NFS endpoint through the Longhorn share manager, so we need to focus on how the downstream cluster interacts with NFS. Once the downstream cluster interacts with this NFS endpoint, we can attach the RWX volume to any workload through the Harvester CSI driver.
+
+### User Stories
+
+#### 1. The Configuration Backup of NeuVector
+NeuVector will backup the configuration files (Policies, Rules, user data, integrations, etc) in the controller. The backup needs the RWX volume to store to ensure this configuration is safe. Currently, we only support RWO volume. The user needs to wait for this feature so that they can use NeuVector in the downstream cluster.
+
+Reference: https://open-docs.neuvector.com/deploying/production/#persistent-volume-example
+
+#### 2. The Backup Store of MongoDB
+MongoDB could configure the backup store with the Kubernetes Operator. The prerequisite is the RWX volume, which is similar to the NeuVector. The user needs this feature once they want to use K8s Operator to backup MongoDB.
+
+Reference: https://www.mongodb.com/docs/kubernetes-operator/v1.25/tutorial/configure-file-store/
+
+### User Experience In Detail
+
+With this feature, the user could create volumes with two options in the downstream cluster. The first is the RWO volume. The second is the RWX volume. Then user could attach use the RWX volume as previous RWO volume. The Harvester CSI driver would handle the internal work.
+The UX did not change too much. The only difference is the user needs to select the RWX volume.
+
+### API changes
+TBD
+
+## Design
+
+### Implementation Overview
+
+A quick overview of the design could be checked below:
+
+![Downstream RWX Support](20240523-support-rwx-volume-to-downstream-cluster/general-rwx.png)
+
+### Solution 1.
+
+Like the RWO solution, we would like to leverage the libvirt capability to hotplug the filesystem and then use the workload pod as a host mountpoint, but currently, it is not supported. The hotplug filesystem on the kubevirt side is referred to as virtiofs. However, virtiofs does not support live migration, so we could not use this solution.
```

Comment:
This should be the general approach if we want to support this for 3rd storage eventually.